src/helper.js

```diff
@@ -53,3 +54,226 @@ export function getOverlays(edge) {
+    visibleEdgesMap.forEach(edge => {
+      visibleVertices.set(
+        edge.sourceId,
+        universalVerticesMap.get(edge.sourceId)
+      );
+      visibleVertices.set(
+        edge.targetId,
+        universalVerticesMap.get(edge.targetId)
+      );
+    });
+
+    return visibleVertices;
+  }
+);
+
+export const getVisibleEdges = memoizeOne(
+  (viewport, xIntervalTree, yIntervalTree, version) => {
+    const xEdgesMap = new Map();
+    const yEdgesMap = new Map();
+    const visibleVertices = new Map();
+    xIntervalTree.queryInterval(
+      viewport.xMin,
+      viewport.xMax,
+      ([low, high, edge]) => {
+        xEdgesMap.set(edge.id, edge);
+      }
+    );
+    yIntervalTree.queryInterval(
+      viewport.yMin,
+      viewport.yMax,
+      ([low, high, edge]) => {
+        yEdgesMap.set(edge.id, edge);
+      }
+    );
+
+    xEdgesMap.forEach((edge, edgeId) => {
+      if (yEdgesMap.has(edgeId)) {
+        visibleVertices.set(edgeId, edge);
+      }
+    });
+
+    return visibleVertices;
+  }
+);
+
+export function addEdge(vToEMap, edge, vertexId) {
+  let sourceVertexEdgeList = vToEMap.get(vertexId);
+  if (sourceVertexEdgeList) {
+    sourceVertexEdgeList.push(edge);
+  } else {
+    vToEMap.set(vertexId, [edge]);
+  }
+}
+
+export function removeEdge(vToEMap, edgeId, vertexId) {
+  let sourceVertexEdgeList = vToEMap.get(vertexId);
+  if (sourceVertexEdgeList) {
+    sourceVertexEdgeList = sourceVertexEdgeList.filter(
+      presentEdge => presentEdge.id !== edgeId
+    );
+  }
+  if (!sourceVertexEdgeList || !sourceVertexEdgeList.length) {
+    vToEMap.delete(vertexId);
+  } else {
+    vToEMap.set(vertexId, sourceVertexEdgeList);
+  }
+}
+
+export const getViewport = memoizeOne(
+  (scrollLeft, scrollTop, clientWidth, clientHeight, zoom) => {
+    const scale = 1 / zoom;
+
+    const scrollLeftScaled = scrollLeft * scale;
+    const clientWidthScaled = clientWidth * scale;
+    const scrollTopScaled = scrollTop * scale;
+    const clientHeightScaled = clientHeight * scale;
+
+    return {
+      xMin: scrollLeftScaled,
+      xMax: scrollLeftScaled + clientWidthScaled,
+      yMin: scrollTopScaled,
+      yMax: scrollTopScaled + clientHeightScaled
+    };
+  }
+);
+
+export function removeNode(intervalTree, intervalTreeNodes, nodeId) {
+  if (!intervalTreeNodes[nodeId]) {
+    return;
+  }
+  intervalTree.remove(intervalTreeNodes[nodeId]);
+  delete intervalTreeNodes[nodeId];
+}
+
+export function makeXIntervalForEdge(edge, v1, v2) {
+  const x1 = Math.min(v1.left, v2.left) || 0;
+  const x2 = Math.max(
+    (v1.left || 0) + (v1.width || 0),
+    (v2.left || 0) + (v2.width || 0)
+  );
+  return [x1, x2, edge];
+}
+
+export function makeYIntervalForEdge(edge, v1, v2) {
+  const y1 = Math.min(v1.top, v2.top) || 0;
+  const y2 = Math.max(
+    (v1.top || 0) + (v1.height || 0),
+    (v2.top || 0) + (v2.height || 0)
+  );
+  return [y1, y2, edge];
+}
+
+export const getTranslate3DCoordinates = (
```

Comment:
can you please add a comment for the underlying logic within `getTranslate3DCoordinates`, `getScaledScrollValues` & `getScaledViewportDimensions`